src/object_store.cpp

```diff
@@ -242,13 +274,22 @@ StringData ObjectStore::get_primary_key_for_object(Group const& group, StringDat
 void ObjectStore::set_primary_key_for_object(Group& group, StringData object_type, StringData primary_key) {
     TableRef table = group.get_table(c_primaryKeyTableName);
 
-    // get row or create if new object and populate
     size_t row = table->find_first_string(c_primaryKeyObjectClassColumnIndex, object_type);
+
+#if REALM_HAVE_SYNC_STABLE_IDS
+    // sync::create_table* functions should have already updated the pk table.
+    if (primary_key.size() && sync::has_object_ids(group)) {
+        REALM_ASSERT(row != not_found);
+        REALM_ASSERT(table->get_string(c_primaryKeyPropertyNameColumnIndex, row) == primary_key);
+    }
```

Comment:
// sync::create_table* functions should have already updated the pk table.
    if (sync::has_object_ids(group)) {
        if (primary_key.empty())
            REALM_ASSERT(row == not_found);
        else {
             REALM_ASSERT(row != not_found);
             REALM_ASSERT(table->get_string(c_primaryKeyPropertyNameColumnIndex, row) == primary_key);
        }
        return;
    }

Both cases seem worth checking, and no need to continue on to do a redundant check in the case where there is a PK.